### PR TITLE
Fix projection name with DataFrame::with_column and window functions

### LINKED ISSUE #11982: Window functions create unwanted projection
### Describe the bug

When using a window function, you get an additional projection that is not expected.

### To Reproduce

```
    let schema = Schema::new(vec![
        Field::new("a", DataType::Int32, true),
    ]);

    let batch = RecordBatch::try_new(
        Arc::new(schema.clone()),
        vec![
            Arc::new(Int32Array::from(vec![1, 2, 3, 4, 5])),
        ],
    )?;

    let ctx = SessionContext::new();

    let provider = MemTable::try_new(Arc::new(schema), vec![vec![batch]])?;
    ctx.register_table("t", Arc::new(provider))?;

    let df = ctx.table("t").await?;

    let func = Expr::WindowFunction(WindowFunction::new(
        WindowFunctionDefinition::BuiltInWindowFunction(BuiltInWindowFunction::RowNumber),
        vec![],
    )).alias("row_num");

    df.clone().select(vec![col("a"), func.clone()])?.show().await?;

    df.with_column("r", func)?.show().await?;
```

### Expected behavior

When using the `select()` operation you get the desired result:

```
+---+---------+
| a | row_num |
+---+---------+
| 1 | 1       |
| 2 | 2       |
| 3 | 3       |
| 4 | 4       |
| 5 | 5       |
+---+---------+
```

But when using `with_column()` you get this additional projection:

```
+---+-----------------------------------------------------------------------+---+
| a | ROW_NUMBER() ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING | r |
+---+-----------------------------------------------------------------------+---+
| 1 | 1                                                                     | 1 |
| 2 | 2                                                                     | 2 |
| 3 | 3                                                                     | 3 |
| 4 | 4                                                                     | 4 |
| 5 | 5                                                                     | 5 |
+---+-----------------------------------------------------------------------+---+
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by devanbenz**
Signed-off-by: Devan <devandbenz@gmail.com>## Which issue does this PR close?

Closes #11982 

## Rationale for this change

Previous usage of dataframe `with_column` using a Window expression was causing an unnecessary projection in the output. 

```
+---+-----------------------------------------------------------------------+---+
| a | ROW_NUMBER() ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING | r |
+---+-----------------------------------------------------------------------+---+
| 1 | 1                                                                     | 1 |
| 2 | 2                                                                     | 2 |
| 3 | 3                                                                     | 3 |
| 4 | 4                                                                     | 4 |
| 5 | 5                                                                     | 5 |
+---+-----------------------------------------------------------------------+---+
```

I also tested manually with a few other `Expr` to test if that would cause issues by filtering the qualifier:

```
    let func = Expr::Literal(ScalarValue::Int32(Some(10)));
    df.with_column("r", func)?.show().await?;
```
outputs:

```
+---+----+
| a | r  |
+---+----+
| 5 | 10 |
| 4 | 10 |
| 3 | 10 |
| 2 | 10 |
| 1 | 10 |
+---+----+
```

### CODE REVIEW

File: datafusion/core/src/dataframe/mod.rs [hunk 1]
```diff
@@ -2869,6 +2876,39 @@ mod tests {
...
+        let df = test_table().await?.select_columns(&["c1", "c2", "c3"])?;
+        let ctx = SessionContext::new();
+        let df_impl = DataFrame::new(ctx.state(), df.plan.clone());
+        let func = Expr::WindowFunction(WindowFunction::new(
+            WindowFunctionDefinition::BuiltInWindowFunction(
+                BuiltInWindowFunction::RowNumber,
+            ),
+            vec![],
+        ))
+        .alias("row_num");
```
Comment (alamb) [inline datafusion/core/src/dataframe/mod.rs:2890]: I think you can use the expr fn here and make this more concise:



```suggestion
        let func = row_number().alias("row_num");
```

Comment (alamb) [reply datafusion/core/src/dataframe/mod.rs:2890]: I also ran this test without the code changes and it fails like this:

```

assertion `left == right` failed
  left: 4
 right: 5

Left:  4
Right: 5
<Click to see difference>

thread 'dataframe::tests::test_window_function_with_column' panicked at datafusion/core/src/dataframe/mod.rs:2882:9:
assertion `left == right` failed
  left: 4
 right: 5
stack backtrace:
   0: rust_begin_unwind
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/std/src/panicking.rs:652:5
...

```

And the output was like

```
[
    "+----+----+-----+-----------------------------------------------------------------------+---+",
    "| c1 | c2 | c3  | ROW_NUMBER() ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING | r |",
    "+----+----+-----+-----------------------------------------------------------------------+---+",
    "| c  | 2  | 1   | 1                                                                     | 1 |",
    "| d  | 5  | -40 | 2                                                                     | 2 |",
    "+----+----+-----+-----------------------------------------------------------------------+---+",
]
```

**Review by alamb — APPROVED** [review]
Looks good to me -- thank you @devanbenz 

cc @timsaucer  thank you for the report. Perhaps you have time to review the change as well

File: datafusion/core/src/dataframe/mod.rs (hunk 1 above)
Comment (devanbenz) [reply datafusion/core/src/dataframe/mod.rs]: Sounds good I went ahead and used the more concise method call. Thanks!

**Review by timsaucer — APPROVED** [review]
LGTM. I also tested locally and it produces the expected results. Thank you for taking care of this!

**Comment by alamb** [comment]
Thanks again @devanbenz  and @timsaucer  -- getting better every day!

**Pull request merged**